qiskit-superstaq/qiskit_superstaq/superstaq_job.py

```diff
@@ -90,35 +128,70 @@ def result(self, timeout: Optional[float] = None, wait: float = 5) -> qiskit.res
             }
         )
 
-    def status(self) -> qiskit.providers.jobstatus.JobStatus:
-        """Query for the job status."""
+    def _check_if_stopped(self) -> None:
+        """Verifies that the job status is not in a stopped state and raises an
+        exception if it is.
 
-        job_id_list = self._job_id.split(",")  # separate aggregated job ids
+        Raises:
+            SuperstaQUnsuccessfulJob: If the job has failed, been canceled, or deleted.
+            SuperstaQException: If unable to get the status of the job from the API.
+        """
+        if self._job_info["status"] in self.STOPPED_STATES:
+            raise gss.superstaq_exceptions.SuperstaQUnsuccessfulJobException(
+                self.get_job_id(), self._job_info["status"]
+            )
 
-        status = "Done"
+    def _refresh_job(self) -> None:
```

Comment:
1. The thought was that `_refresh_job()` would not be called by the user (they would call `status()` and `_refresh_job()` is only ever used in `status()`) so the check for a terminal state is first done in `status()` and then only is `_refresh_job()` called. But I can move that check into `_refresh_job()` if it'd be better?
2/3. I've now stored each job status in `self._job_info[job_id]` and gone for an overall status that is determined by the "worst status" of all the jobs.
4. I think I'd prefer to leave it for another PR for now

Reply:
1. The thought was that `_refresh_job()` would not be called by the user (they would call `status()` and `_refresh_job()` is only ever used in `status()`) so the check for a terminal state is first done in `status()` and then only is `_refresh_job()` called. But I can move that check into `_refresh_job()` if it'd be better?
2. I've now stored each job status in `self._job_info[job_id]` and gone for an overall status that is determined by the "worst status" of all the jobs.
3. Will implement into `_wait_for_results` if 2. looks good
4. I think I'd prefer to leave it for another PR for now

Reply:
1. ah ok, hadn't noticed that it wasn't called anywhere else. i think whichever way you prefer is fine then
2. and 3. i was thinking that the entirety of the `result` dictionary should be stored, instead of just `result["status"]`. That way once a job is done we won't need to query the server about it again to get its results, and e.g. `_wait_for_results` can just be something like
```python
def _wait_for_results(self, timeout, wait):
    self.wait_for_final_state(timeout, wait)  # calls status() until done
    self._check_if_stopped()
    return [self._job_info[job_id] for job_id in self._job_id.split(",")]
```
4. sounds good!

Reply:
That's a good idea, I have now stored the whole result in `self._job_info` (see 3c9aed3). The server is only queried within `_refresh_job()` and `refresh_job()` is only called by `status()`